exercises/5.1/template_ws/ros2/lesson_perception/src/perception_node.cpp

```diff
@@ -0,0 +1,211 @@
+
+
+            /* ========================================
+             * Fill Code: PASSTHROUGH FILTER(S)
+             * ========================================*/
+
+
+            /* ========================================
+             * Fill Code: CROPBOX (Optional)
+             * ========================================*/
+
+
+            /* ========================================
+             * Fill Code: PLANE SEGEMENTATION
+             * ========================================*/
+
+
+
+            /* ========================================
+             * Fill Code: EUCLIDEAN CLUSTER EXTRACTION
+             * ========================================*/
+
+
+
+            /* ========================================
+             * Fill Code: STATISTICAL OUTLIER REMOVAL
+             * ========================================*/
+
+
+            /* ========================================
+             * Fill Code: PUBLISH OTHER MARKERS (OPTIONAL)
+             * ========================================*/
+
+
+            /* ========================================
+             * BROADCAST TRANSFORM
+             * ========================================*/
+
+
+            /* ========================================
+             * Fill Code: POLYGONAL SEGMENTATION
+             * ========================================*/
+
+
+            /* ========================================
+             * CONVERT PointCloud2 PCL->ROS
+             * PUBLISH CLOUD
+             * Fill Code: UPDATE AS NECESSARY
+             * ========================================*/
+
+            this->publishPointCloud(voxel_grid_publisher_, *cloud);
+
+        }
+
+        void publishPointCloud(rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr publisher,
+                               pcl::PointCloud<pcl::PointXYZ> point_cloud) {
+
+            sensor_msgs::msg::PointCloud2::SharedPtr pc2_cloud(new sensor_msgs::msg::PointCloud2);
+            pcl::toROSMsg(point_cloud, *pc2_cloud);
+            pc2_cloud->header.frame_id = world_frame;
+            pc2_cloud->header.stamp = this->get_clock()->now();
+            publisher->publish(*pc2_cloud);
+        }
+
+        rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_subscriber_;
+
+        /*
+         * Publishers
+         */
+        rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr voxel_grid_publisher_;
+
+
+        /*
+         * Parameters
+         */
+        std::string cloud_topic;
+        std::string world_frame;
+        std::string camera_frame;
+
+        /*
+         * TF
+         */
+        std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
+        std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+        std::unique_ptr<tf2_ros::TransformBroadcaster> br;
+};
+
+
+int main(int argc, char *argv[])
+{
+    /*
+     * INITIALIZE ROS NODE
+     */
+    rclcpp::init(argc, argv);
+    rclcpp::spin(std::make_shared<PerceptionNode>());
```

Comment:
Oof I do not like this line.  Is it technically correct?  Yes.  Should it work?  Yes.  Flawlessly?  Yes.  Do I like it?  No.

What's so bad about just spending the one extra line to construct the thing and then passing that thing in?

This only just barely works, because it will get copied, which increments the number of shared pointers to 2, mere clock cycles before the one `make_shared<>()` constructed is destructed.